enhancements/cluster-logging/elasticsearch-v2-api-and-infrastructure.md

```diff
@@ -0,0 +1,300 @@
+
+1. Create their elasticsearch CR specifying a particular sized cluster (based on t-shirt sizes)
+
+Current specification:
+
+```yaml
+apiVersion: "logging.openshift.io/v1"
+kind: "Elasticsearch"
+metadata:
+  name: "elasticsearch"
+spec:
+  managementState: "Managed"
+  nodeSpec:
+    resources:
+      limits:
+        memory: 1Gi
+      requests:
+        cpu: 100m
+        memory: 1Gi
+  nodes:
+  - nodeCount: 1
+    roles:
+    - client
+    - data
+    - master
+    storage:
+      storageClassName: gp2
+      size: 10Gi
+  redundancyPolicy: ZeroRedundancy
+  indexManagement:
+    policies:
+    - name: infra-policy
+      pollInterval: 1m
+      phases:
+        hot:
+          actions:
+            rollover:
+              maxAge:   2m
+        delete:
+          minAge: 5m
+    mappings:
+    - name:  infra
+      policyRef: infra-policy
+      aliases:
+      - infra
+      - logs.infra
+```
+
+Proposed:
+
+```yaml
+apiVersion: "logging.openshift.io/v1"
+kind: "Elasticsearch"
+metadata:
+  name: "elasticsearch"
+spec:
+  storage:
+    storageClassName: gp2
+    size: 10Gi
+  size: 0x.medium
+  indexManagement:
+    infraPolicies:
+    - phases:
+      hot:
+        actions:
+          rollover:
+            maxAge:   2m
+      delete:
+        minAge: 5m
+    appPolicies:
+    - phases:
+      hot:
+        actions:
+          rollover:
+            maxAge: 2m
+      delete:
+        minAge: 5m
+```
+
+## Proposal
+
+Currently the Elasticsearch CR is overly complicated and allows for too much granular control over the data nodes, which places the expectation
+of expertise on the user rather than containing it to the operator (which is where it should reside). This proposal seeks to outline a v2 to provide
+a more simple means of interracting with the Elasticsearch Operator to ask for a managed Elasticsearch cluster that can be better
+configured for performance.
+
+### User Stories
+
+#### As an user of EO, I want to have to provide minimal information for my cluster to be created for me
+
+As an user I want to just be able to provide information such as a t-shirt size (which is outlined further for me) and my storage information
+and have an elasticsearch cluster spun up for me so that I do not need to be an expert and configure my node roles and node resources.
+
+### Implementation Details
+
+#### Assumptions
+
+* Currently the EO CRD version is V1 and this would be a V2. We would assume that EO is able to react for both a V1 and V2 and if the user creates a V2
+  elasticsearch CR with the same name as an existing V1, the operator would migrate the V1 backing storage naming to be used for V2 (read: it would
+  rename the PVCs and rebind the backing PV so that it matches the updated naming convention).
+
+### Risks and Mitigations
+
+## Not Breaking Other Operators who Depend on EO
+
+As part of adding another version of the CRD that EO will respond to, we need to ensure that we can continue to support those that remain on V1 so that
```

Comment:
those usually come in the form of a transformation hook though right? I'll spell it out a bit more in the doc but my thought was this:
If there exists a V1 CR named "elasticsearch" in the namespace "openshift-logging"
and then a V2 CR named "elasticsearch" in the namespace "openshift-logging" is created
The operator would ignore the V1 one and migrate the objects from V1 to V2 (objects primarily being currently used storage)

My main concern with doing a straight migration is that we would need to try to extrapolate the desired t-shirt size based on the current configurations which may be very difficult to get correct. It may also be that the customer's current configuration does not correctly meet their demands so doing a migration could leave them with similar issues (mainly if their storage capacity isn't high enough or if their cluster is under resourced)

Reply:
Yep, it's a trade off: the migration hook is a "standard" way of doing this and allows you to provide a v2-only operator, but it requires that an automatic v1->v2 translation will not break existing v1 deployments.